### PR TITLE
types, expression: fix convert string to int

### LINKED ISSUE #7511: BIT_XOR get different result from mysql
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
```
CREATE TABLE `test`.`testxor` (
  `a` varchar(13) COLLATE utf8_unicode_ci NOT NULL,
  `b` varchar(13) COLLATE utf8_unicode_ci NOT NULL,
  `c` varchar(38) COLLATE utf8_unicode_ci NOT NULL,
  `d` varchar(1) COLLATE utf8_unicode_ci NOT NULL,
  `e` varchar(13) COLLATE utf8_unicode_ci NOT NULL,
  `f` varchar(2) COLLATE utf8_unicode_ci DEFAULT NULL,
  `g` varchar(2) COLLATE utf8_unicode_ci DEFAULT NULL,
  `h` varchar(31) COLLATE utf8_unicode_ci DEFAULT NULL,
  `i` varchar(3) COLLATE utf8_unicode_ci DEFAULT NULL,
  `j` float DEFAULT NULL,
  `k` varchar(4) COLLATE utf8_unicode_ci DEFAULT NULL,
  `l` varchar(2) COLLATE utf8_unicode_ci DEFAULT NULL,
  `m` varchar(14) COLLATE utf8_unicode_ci DEFAULT NULL,
  `n` varchar(14) COLLATE utf8_unicode_ci DEFAULT NULL,
  `o` varchar(64) COLLATE utf8_unicode_ci DEFAULT NULL,
  `p` varchar(14) COLLATE utf8_unicode_ci DEFAULT NULL,
  `q` varchar(14) COLLATE utf8_unicode_ci DEFAULT NULL,
  `r` varchar(64) COLLATE utf8_unicode_ci DEFAULT NULL,
  `s` varchar(14) COLLATE utf8_unicode_ci DEFAULT NULL,
  `t` varchar(128) COLLATE utf8_unicode_ci DEFAULT NULL,
  `u` varchar(1) COLLATE utf8_unicode_ci DEFAULT NULL,
  UNIQUE KEY `key1` (`a`,`b`,`c`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;

insert into test.testxor values("IRwNVi0VrwPTz","8HxZlIXM2pqXl","4pOdkHBSBe3eq9aAIvvRKy","W","FYwb","U","iE","1W40","Yeo",275632000,"XQ","e","1W40","XZs","RvWGWvySa05WJT5gy7bs9wYgEwDYBUmAVHChSz0w","AR","71LXid7pNS","9kXtYca3vHqcV","2z","dGIKm1zja6COUmzcIYYTAV5DGsyenKi3sXGT8UJMePKc9t4hENo2Pj97SvmdLXiqvkPajBoK7OWr","0");

insert into test.testxor values("8E2kS4k","m0EnLCHM","JvnlGJvXhqXOSY4FUnbE","Z","yykhS","m","s","3GL5","30y",275620000,"FFg","GI","3GL5","bggqNaXvI0Q8u", "8Q5JGMXRZd6Zhwak69PPdTeTi3Iz37fOz04dZK4Fyi3syZxu","iy3o3VG","CAL0yKLgT","jsBxc4ahO95lI9W7yCjL5A2JipozStcize2j","K27L","EttRkgKjm0j7RCs29U","5");
```

2. What did you expect to see?
execute in mysql:
```
mysql> SELECT BIT_XOR(CONCAT_WS(',',a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u)) AS result FROM test.testxor;
+--------+
| result |
+--------+
|      8 |
+--------+
1 row in set (0.00 sec)
```
I want see the same result when execute in tidb.

3. What did you see instead?
execute in tidb:
```
mysql> SELECT BIT_XOR(CONCAT_WS(',',a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u)) AS result FROM test.testxor;
+--------+
| result |
+--------+
|    800 |
+--------+
1 row in set, 2 warnings (0.01 sec)
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
tidb version:
```
tidb_version(): Release Version: v2.1.0-rc.1-8-g55565f1
Git Commit Hash: 55565f1c938f47ab09bae8b04dc4edc29ce4d8e3
Git Branch: master
UTC Build Time: 2018-08-28 04:09:51
GoVersion: go version go1.10.3 darwin/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
```
mysql version:  5.7.18

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve? 
fix #7511 

### What is changed and how it works?
refine `getValidIntPrefix` which call `getValidFloatPrefix` before.

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported function/method change

Side effects

Related changes

 - Need to cherry-pick to the release branch

release-2.0 release-2.1

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8157)

### CODE REVIEW

**Comment by winoros** [comment]
@XuHuaiyu CI failed.

File: types/convert.go [hunk 1]
```diff
@@ -231,12 +231,26 @@ func NumberToDuration(number int64, fsp int) (Duration, error) {
...
 // getValidIntPrefix gets prefix of the string which can be successfully parsed as int.
-func getValidIntPrefix(sc *stmtctx.StatementContext, str string) (string, error) {
-	floatPrefix, err := getValidFloatPrefix(sc, str)
-	if err != nil {
-		return floatPrefix, errors.Trace(err)
+func getValidIntPrefix(sc *stmtctx.StatementContext, str string) (valid string, err error) {
+	var validLen int
+	for i := 0; i < len(str); i++ {
+		c := str[i]
+		if c == '+' || c == '-' {
```
Comment (lamxTyler) [inline types/convert.go]: What about `select bit_xor('++8e2')`?

**Comment by lzmhhh123** [comment]
We should consider the string like **_1.1e-6_**. This bug may be fixed by #8279 .

File: types/convert.go [hunk 2]
```diff
@@ -231,12 +231,26 @@ func NumberToDuration(number int64, fsp int) (Duration, error) {
...
+			validLen++
+		} else if c < '0' || c > '9' {
```
Comment (lzmhhh123) [inline types/convert.go]: The string like _**1.1e-6**_ should be considered.

**Comment by XuHuaiyu** [comment]
As we discussed in this https://github.com/pingcap/tidb/issues/9905, this is acceptable behavior.

**Pull request closed without merging**